keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,237 @@
+[2] [Gc orphaned pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L97)
+[3] [requirements on node objects](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L167)
+[8] [bootstrap checkpointing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/0004-bootstrap-checkpointing.md)
+[9] [bootstrap checkpointing deprecation intent](https://github.com/kubernetes/enhancements/issues/1103)
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Introduce a new API to report node state
+* Handle cases where the Kubelet itself is unresponsive
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP do not intend to introduce any new API, instead we plan on re-using the Lease API and a mechanism which is similar to leader election (if not the same).
+As stated this KEP is all about finding the right inter-lock between all the actors. The main are:
+
+* Kube-Controller-Manager
+* Kubelets
+
+
+
+#### kube-controller-manager
+
+
+kube-controller-manager needs to provide a controller that:
+
+Calls the cloud provider
+Apply the node-shutdown taint
+Executes the detach operation
+Removes the taint
+
+given the current design of our controllers and their following of the single-responsibility pattern, we’ll introduce a new controller called gc_volume_controller under pkg/controller/gc, we will also make the attach_detach_controller skip shutdown nodes to avoid concurrent detach requests.
+
+This gc_volume_controller watches node objects and when notified (each step describe the possible race condition that could happen)
+
+1. checks the node state via IsShutdownByProviderID(): if there’s a race condition with the cloud provider (i.e. the instance came back see 2.a)
+2. apply a node-shutdown taint
+3. tries to hold node heartbeat lease
+  3.a. fails: backoff and retry if max retries hit -> aborts
+  3.b. success: hold the lease for this node: if the node comes back at this moment it won’t be able to acquire the lease -> pods won’t start
+4. register node as being processed in order to renew the Lease periodically
+5. delete the pod with a 0-grace-period
+6. detach the volumes
+7. once 6. returns untaint the node and stop renewing the Lease
+
+
+To avoid being too reactive in 2. we can backoff and not try to acquire the lock immediately to avoid evicting too fast
+
+#### Kubelet
+
+we need also to apply changes to the kubelet, at startup the kubelet will:
+
+1. acquire the Lease:
+  a. fails: do not start any pods, unless the pod is:
+     i. Tolerate the node shutdown taint
+     ii. Is a static Pod 
```

Comment:
Is there a reason why the static pod wouldn't just tolerate the taint?  Why make this exception?

Reply:
when upgrading what will happen to static pods that weren't tolerating the taint (e.g. the apiserver, scheduler etc..) ?

Reply:
We need to support a single version of skew between the kubelet and master components based on https://kubernetes.io/docs/setup/release/version-skew-policy/, since the kubelet could be the same version as the apiserver, and other master components can be 1 version behind the apiserver.

I'm having a hard time thinking of a scenario in which static pods would need the taint.  During upgrades, the master node should not have the shutdown (or whatever we name it) taint.  If the node is being shut down, I don't think running all static pods (other than the apiserver, which must be ahead of the kubelet version) is necessary.

Alternatively, we can add the taint in version N (and add tolerations to static pods), and then add the kubelet behavior based on the taint in version N+1.

Reply:
> We need to support a single version of skew between the kubelet and master components

https://kubernetes.io/docs/setup/release/version-skew-policy/#kubelet allows kubelet to be up to 2 versions older than the apiserver and other master components

Reply:
Sorry, I should specify that the kubelet can be at most 1 version _**ahead of**_ a master component.

Reply:
I'm fine with adding an `action required` on the release note that saying that to start static pods  when a node is shutdown they should add the corresponding toleration, wdyt @dashpole @liggitt ?

Reply:
That sounds reasonable to me, as I don't believe adding a new taint/toleration such as this is backwards-incompatible.